lib/puppet/util/execution.rb

```diff
@@ -179,8 +179,19 @@ def self.execute(command, options = NoOptionsSpecified)
       if execution_stub = Puppet::Util::ExecutionStub.current_value
         return execution_stub.call(*exec_args)
       elsif Puppet.features.posix?
-        child_pid = execute_posix(*exec_args)
-        exit_status = Process.waitpid2(child_pid).last.exitstatus
+        begin
+          child_pid = execute_posix(*exec_args)
+          exit_status = Process.waitpid2(child_pid).last.exitstatus
+          child_pid = nil
+        rescue Timeout::Error => e
```

Comment:
An alternative approach would be to use an `ensure` block, but it seems a bit heavy-handed.